eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -312,11 +315,13 @@ processDelegationInternal preserveIncorrectDelegation stakeCred mCurDelegatee ne
       cState
         & certDStateL . accountsL
           %~ adjustAccountState (stakePoolDelegationAccountStateL ?~ stakePool) stakeCred
+        & certPStateL %~ adjustPState stakePool
     delegVote dRep cState =
       let cState' =
-            processDRepUnDelegation stakeCred mCurDelegatee cState
+            cState
               & certDStateL . accountsL
                 %~ adjustAccountState (dRepDelegationAccountStateL ?~ dRep) stakeCred
+              & certVStateL %~ unDelegDRep stakeCred mCurDelegatee
           dReps
             | preserveIncorrectDelegation = cState ^. certVStateL . vsDRepsL
             | otherwise = cState' ^. certVStateL . vsDRepsL
```

Comment:
This will need to be slightly rewritten:
```suggestion
      in if preserveIncorrectDelegation
         then cState & certVStateL %~ case dRep of 
                DRepCredential dRepCred ->
                  vsDRepsL %~ Map.adjust (drepDelegsL %~ Set.insert stakeCred) dRepCred
                _ -> id
         else cState'
```
And all the stuff below is no longer needed.